### ISSUE TITLE
Random AWS delays can cause Travis failure

### ISSUE THREAD

**Issue by ljbade** [mapbox/mapbox-gl-native#857]
Due to a flaw in the AWS API I can't set an instances name at the same time I create it. Since we don't want un-named instances running I set the name straight away in a seperate command.

However sometimes AWS is a bit slow and complains that the instance *it just gave me the ID of* does not exist.

I tried adding a delay to fix it, but it only reduced the rate of failures. Example failure [here](https://travis-ci.org/mapbox/mapbox-gl-native/jobs/50316209).

I need a better way of doing this.

/cc @jfirebaugh @freenerd

**Comment by jfirebaugh** [comment]
[Code here](https://github.com/mapbox/mapbox-gl-native/blob/d1102b5b21c88537a8d7f1a94c121dee0243282a/android/scripts/run-build.sh#L48-L61)

**Comment by jfirebaugh** [comment]
@ljbade The error in the build above sounds like it comes from the `run-instances` command, not the `create-tags` command:

"A client error (AuthFailure) occurred when calling the RunInstances operation: AWS was not able to validate the provided access credentials"

**Comment by freenerd** [comment]
@jfirebaugh yes. @ljbade got another travis example of it failing?

@ljbade instead of doing one sleep and then fail, continue to wait until it's there (probably with a timeout) similar to [how atlas does it](https://github.com/mapbox/atlas-server/blob/mb-pages/dev/ami/build-ami.sh#L73-L78)

**Comment by ljbade** [comment]
@freenerd Hmm I don't think I do currently since I usually just restart the job if it errors, will grab it when I see it again. But usually the `create-tags` command returns an error that the given EC2 instance ID does not exist (which comes from the previous `run-instances` command).

@jfirebaugh I have never seen that error before then. The credentials are the same ones that gl-native has had since forever for uploading test results to s3.

**Comment by ljbade** [comment]
@freenerd Hmm I am not sure to check for when an instance ID becomes valid. I don't want to wait for the instance to be ready before I name it since it takes several minutes for the instance to boot.

**Comment by ljbade** [comment]
@freenerd I am actually thinking this is more an AWS bug, run-instances should not return until the instance ID is valid.

**Comment by freenerd** [comment]
> @freenerd I am actually thinking this is more an AWS bug, run-instances should not return until the instance ID is valid.

Welcome to the beauty of distributed systems.  I'm very sure that `run-instances` doesn't have knowledge if the instance id is replicated everywhere, so it can't wait.

> But usually the create-tags command returns an error that the given EC2 instance ID does not exist (which comes from the previous run-instances command).

@ljbade do something like [in this gist](https://gist.github.com/freenerd/9db25e8e3c6d2e3c617d) (might not be full-compatible bash)

**Comment by ljbade** [comment]
I think I will look into moving to @yhahn's https://github.com/mapbox/cfn-ci

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by ljbade** [comment]
@kkaefer, where did you get with using the Travis Android instance?

**Comment by kkaefer** [comment]
It works: https://travis-ci.org/mapbox/mapbox-gl-native/jobs/51279115, but doesn't use the Android instances (the `language` type in `.travis.yml` is not a matrix option, you only get to choose one for all builds...), so has to download the SDK too. It installs the NDK with mason. We also get some [random download failures](https://travis-ci.org/mapbox/mapbox-gl-native/jobs/51279116#L1048). Some builds [failed due to an overly long log](https://travis-ci.org/mapbox/mapbox-gl-native/jobs/51279119#L760), but we can solve that by disabling the CURL progress bars in mason. [One failure](https://travis-ci.org/mapbox/mapbox-gl-native/jobs/51279120#L1454) is due to mason picking up the wrong env variable.

So in short, we should be able to get this running on all Travis boxes, if we can figure out why the SDK download fails some time. We could also install the SDK with Mason (or cache it on S3 manually, since caches aren't enabled for non-sudo, public builds). If we can get all packages whitelisted, we could also try moving to the Docker build machines and could cache the SDK automatically.

**Comment by jfirebaugh** [comment]
Not using AWS anymore so not applicable.
